kernel/sql-federation/optimizer/src/main/java/org/apache/shardingsphere/sqlfederation/optimizer/converter/segment/projection/impl/DataTypeConverter.java

```diff
@@ -45,7 +45,12 @@ public static Optional<SqlNode> convert(final DataTypeSegment segment) {
         if (null == segment) {
             return Optional.empty();
         }
-        return Optional.of(new SqlDataTypeSpec(new SqlBasicTypeNameSpec(Objects.requireNonNull(SqlTypeName.get(segment.getDataTypeName())), segment.getDataLength().getPrecision(), SqlParserPos.ZERO),
-                SqlParserPos.ZERO));
+        return Optional.of(new SqlDataTypeSpec(getSqlBasicTypeNameSpec(segment), SqlParserPos.ZERO));
+    }
+    
+    private static SqlBasicTypeNameSpec getSqlBasicTypeNameSpec(final DataTypeSegment segment) {
```

Comment:
Hi @zihaoAK47, can you add e2e test case to cover this logic change?

Reply:
Hi mentor, the issue arises from executing the cast function in a federated query scenario. I tried to add the following statement to the e2e process:
"select cast(1 as char)"
This SQL statement will result in incorrect column types error in the JDBC scenario, causing the e2e process to fail.
﻿
I tried the following other SQL statements, but the result obtained from executing e2e is different from the function executed on MySQL:
select cast("2024-03-28" as date)
Expected :2024-03-28
Actual   :19810